python/ray/serve/tests/unit/test_deployment_state.py

```diff
@@ -2372,6 +2348,146 @@ def test_autoscale(mock_deployment_state_manager_full, target_capacity_direction
+    mock_deployment_state_manager_full,
+):
+    """This test asserts that when downscaling first any non-running replicas are
+    scavenged, before stopping fully running replicas
+
+    More context on the issue could be found in:
+    https://github.com/ray-project/ray/issues/43034
+    """
+
+    app_name = "test_app"
+    deployment_name = "deployment_with_slow_to_start_replicas"
+
+    deployment_id = DeploymentID(deployment_name, app_name)
+
+    # Create deployment state manager
+    create_deployment_state_manager, timer, _ = mock_deployment_state_manager_full
+    dsm: DeploymentStateManager = create_deployment_state_manager()
+
+    # Deploy deployment with 3 replicas
+    info, _ = deployment_info(
+        autoscaling_config={
+            "target_num_ongoing_requests_per_replica": 1,
+            "min_replicas": 0,
+            "max_replicas": 6,
+            "initial_replicas": 3,
+            "upscale_delay_s": 0,
+            "downscale_delay_s": 0,
+        }
+    )
+
+    dsm.deploy(deployment_id, info)
+
+    deployment_state: DeploymentState = dsm._deployment_states[deployment_id]
+
+    # status=UPDATING, status_trigger=DEPLOY
+    dsm.update()
+    check_counts(deployment_state, total=3, by_state=[(ReplicaState.STARTING, 3)])
+    assert deployment_state.curr_status_info.status == DeploymentStatus.UPDATING
+    assert (
+        deployment_state.curr_status_info.status_trigger
+        == DeploymentStatusTrigger.CONFIG_UPDATE_STARTED
+    )
+
+    # Set replicas as SUCCESSFUL and check statuses
+    for replica in deployment_state._replicas.get():
+        replica._actor.set_ready()
+
+    # status=HEALTHY, status_trigger=DEPLOY
+    dsm.update()
+    check_counts(deployment_state, total=3, by_state=[(ReplicaState.RUNNING, 3)])
+    assert deployment_state.curr_status_info.status == DeploymentStatus.HEALTHY
+    assert (
+        deployment_state.curr_status_info.status_trigger
+        == DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED
+    )
+
+    # Fetch all currently running replicas
+    running_replicas = deployment_state._replicas.get(states=[ReplicaState.RUNNING])
+
+    for replica in deployment_state._replicas.get():
+        dsm.record_autoscaling_metrics(replica._actor.replica_tag, 2, timer.time())
+
+    # status=UPSCALING, status_trigger=AUTOSCALE
+    dsm.update()
+    check_counts(
+        deployment_state,
+        total=6,
+        by_state=[(ReplicaState.RUNNING, 3), (ReplicaState.STARTING, 3)],
+    )
+    assert deployment_state.curr_status_info.status == DeploymentStatus.UPSCALING
+    assert (
+        deployment_state.curr_status_info.status_trigger
+        == DeploymentStatusTrigger.AUTOSCALING
+    )
+
+    # Set replicas as PENDING_INITIALIZATION: actors have been successfully allocated,
+    # but replicas are still pending successful initialization
+    for replica in deployment_state._replicas.get():
+        replica._actor.set_pending_init()
```

Comment:
can we please also test the case for `PENDING_ALLOCATION`? you could just parametrize the test for each case

(in the future, we should also verify that `PENDING_ALLOCATION` is preferred over `PENDING_INITIALIZATION`)